### ISSUE TITLE
Some moves with disambiguators cannot cause check/checkmate

### ISSUE THREAD

**Issue by jacksonthall22** [jacksonthall22/SAN-strings#4]
From https://chess.stackexchange.com/a/45638/19286:

> Even if a disambiguated move is itself possible, it may be incompatible with checks/checkmates. Any move that can't possibly cause new squares to become attacked cannot cause check or checkmate. This is relevant if disambiguation in the move happens to make the start square and end square badly aligned with other pieces of the same type that must have been on certain squares, especially for queens. For example, Qb2a1 cannot cause new attacks because other queens must have been on b1 and a2 as well, so Qb2a1+ and Qb2a1# are impossible; same for Qc3(a1/b2),Qd4(a1/b2), and maybe more that I haven't found systematically. Edit: I'm not sure about bishops, as eg. Bc3b2 doesn't make the bishops attack more squares but may discover check/mate.

**Comment by phoenizboi** [comment]
Bishop moves can be impossible. An example is Ba1c3+
![IMG_0565](https://github.com/user-attachments/assets/c78c8087-8d10-4ba8-aee8-7bbf8636ac61)
The bishop can't make a discovery attack because it started off in the corner
I'm sorry I can't do much else to help, but I can try and filter out the impossible moves manually

**Comment by phoenizboi** [comment]
I think I got everything. I triple checked it and it looks good, but please check it yourself. Here is every invalid move that should be removed:

<details>
  <summary>Click to expand!</summary>
"Ba1b2+", "Ba1b2#", "Ba1c3+", "Ba1c3#", "Ba1d4+", "Ba1d4#", "Ba8b7+", "Ba8b7#", "Ba8c6+", "Ba8c6#", "Ba8d5+", "Ba8d5#", "Bh1e4+", "Bh1e4#", "Bh1f3+", "Bh1f3#", "Bh1g2+", "Bh1g2#", "Bh8e5+", "Bh8e5#", "Bh8f6+", "Bh8f6#", "Bh8g7+", "Bh8g7#", "Qa1h8+", "Qa1h8#", "Qa8h1+", "Qa8h1#", "Qb2a1+", "Qb2a1#", "Qb2h8+", "Qb2h8#", "Qb7a8+", "Qb7a8#", "Qb7h1+", "Qb7h1#", "Qc3a1+", "Qc3a1#", "Qc3h8+", "Qc3h8#", "Qc4a2+", "Qc4a2#", "Qc5a7+", "Qc5a7#", "Qc6a8+", "Qc6a8#", "Qc6h1+", "Qc6h1#", "Qd3b1+", "Qd3b1#", "Qd4a1+", "Qd4a1#", "Qd4b2+", "Qd4b2#", "Qd4h8+", "Qd4h8#", "Qd5a8+", "Qd5a8#", "Qd5b7+", "Qd5b7#", "Qd5h1+", "Qd5h1#", "Qd6b8+", "Qd6b8#", "Qe3g1+", "Qe3g1#", "Qe4a8+", "Qe4a8#", "Qe4g2+", "Qe4g2#", "Qe4h1+", "Qe4h1#", "Qe5a1+", "Qe5a1#", "Qe5g7+", "Qe5g7#", "Qe5h8+", "Qe5h8#", "Qe6g8+", "Qe6g8#", "Qf3a8+", "Qf3a8#", "Qf3h1+", "Qf3h1#", "Qf4h2+", "Qf4h2#", "Qf5h7+", "Qf5h7#", "Qf6a1+", "Qf6a1#", "Qf6h8+", "Qf6h8#", "Qg2a8+", "Qg2a8#", "Qg2h1+", "Qg2h1#", "Qg7a1+", "Qg7a1#", "Qg7h8+", "Qg7h8#", "Qh1a8+", "Qh1a8#", "Qh8a1+", "Qh8a1#"
</details>

**Comment by phoenizboi** [comment]
Looks like I forgot the moves with captures, here's actually everything:

<details>
  <summary>Click to expand!</summary>
"Ba1b2+", "Ba1b2#", "Ba1c3+", "Ba1c3#", "Ba1d4+", "Ba1d4#", "Ba8b7+", "Ba8b7#", "Ba8c6+", "Ba8c6#", "Ba8d5+", "Ba8d5#", "Bh1e4+", "Bh1e4#", "Bh1f3+", "Bh1f3#", "Bh1g2+", "Bh1g2#", "Bh8e5+", "Bh8e5#", "Bh8f6+", "Bh8f6#", "Bh8g7+", "Bh8g7#", "Qa1h8+", "Qa1h8#", "Qa8h1+", "Qa8h1#", "Qb2a1+", "Qb2a1#", "Qb2h8+", "Qb2h8#", "Qb7a8+", "Qb7a8#", "Qb7h1+", "Qb7h1#", "Qc3a1+", "Qc3a1#", "Qc3h8+", "Qc3h8#", "Qc4a2+", "Qc4a2#", "Qc5a7+", "Qc5a7#", "Qc6a8+", "Qc6a8#", "Qc6h1+", "Qc6h1#", "Qd3b1+", "Qd3b1#", "Qd4a1+", "Qd4a1#", "Qd4b2+", "Qd4b2#", "Qd4h8+", "Qd4h8#", "Qd5a8+", "Qd5a8#", "Qd5b7+", "Qd5b7#", "Qd5h1+", "Qd5h1#", "Qd6b8+", "Qd6b8#", "Qe3g1+", "Qe3g1#", "Qe4a8+", "Qe4a8#", "Qe4g2+", "Qe4g2#", "Qe4h1+", "Qe4h1#", "Qe5a1+", "Qe5a1#", "Qe5g7+", "Qe5g7#", "Qe5h8+", "Qe5h8#", "Qe6g8+", "Qe6g8#", "Qf3a8+", "Qf3a8#", "Qf3h1+", "Qf3h1#", "Qf4h2+", "Qf4h2#", "Qf5h7+", "Qf5h7#", "Qf6a1+", "Qf6a1#", "Qf6h8+", "Qf6h8#", "Qg2a8+", "Qg2a8#", "Qg2h1+", "Qg2h1#", "Qg7a1+", "Qg7a1#", "Qg7h8+", "Qg7h8#", "Qh1a8+", "Qh1a8#", "Qh8a1+", "Qh8a1#", "Ba1xb2+", "Ba1xb2#", "Ba1xc3+", "Ba1xc3#", "Ba1xd4+", "Ba1xd4#", "Ba8xb7+", "Ba8xb7#", "Ba8xc6+", "Ba8xc6#", "Ba8xd5+", "Ba8xd5#", "Bh1xe4+", "Bh1xe4#", "Bh1xf3+", "Bh1xf3#", "Bh1xg2+", "Bh1xg2#", "Bh8xe5+", "Bh8xe5#", "Bh8xf6+", "Bh8xf6#", "Bh8xg7+", "Bh8xg7#", "Qa1xh8+", "Qa1xh8#", "Qa8xh1+", "Qa8xh1#", "Qb2xa1+", "Qb2xa1#", "Qb2xh8+", "Qb2xh8#", "Qb7xa8+", "Qb7xa8#", "Qb7xh1+", "Qb7xh1#", "Qc3xa1+", "Qc3xa1#", "Qc3xh8+", "Qc3xh8#", "Qc4xa2+", "Qc4xa2#", "Qc5xa7+", "Qc5xa7#", "Qc6xa8+", "Qc6xa8#", "Qc6xh1+", "Qc6xh1#", "Qd3xb1+", "Qd3xb1#", "Qd4xa1+", "Qd4xa1#", "Qd4xb2+", "Qd4xb2#", "Qd4xh8+", "Qd4xh8#", "Qd5xa8+", "Qd5xa8#", "Qd5xb7+", "Qd5xb7#", "Qd5xh1+", "Qd5xh1#", "Qd6xb8+", "Qd6xb8#", "Qe3xg1+", "Qe3xg1#", "Qe4xa8+", "Qe4xa8#", "Qe4xg2+", "Qe4xg2#", "Qe4xh1+", "Qe4xh1#", "Qe5xa1+", "Qe5xa1#", "Qe5xg7+", "Qe5xg7#", "Qe5xh8+", "Qe5xh8#", "Qe6xg8+", "Qe6xg8#", "Qf3xa8+", "Qf3xa8#", "Qf3xh1+", "Qf3xh1#", "Qf4xh2+", "Qf4xh2#", "Qf5xh7+", "Qf5xh7#", "Qf6xa1+", "Qf6xa1#", "Qf6xh8+", "Qf6xh8#", "Qg2xa8+", "Qg2xa8#", "Qg2xh1+", "Qg2xh1#", "Qg7xa1+", "Qg7xa1#", "Qg7xh8+", "Qg7xh8#", "Qh1xa8+", "Qh1xa8#", "Qh8xa1+", "Qh8xa1#"
</details>

**Comment by jacksonthall22** [comment]
Hey, thanks! Can you share the code you used to generate this? That logic is ultimately how we should decide if the list is right. Or feel free to fork and submit a PR.

**Comment by phoenizboi** [comment]
Ok im actually completely new to github and don't know how to write programs. I did this all manually. Also i discovered that i messed it up once again, because i found out that captures can allow for a piece to block, and then be captured to reveal a check, making some moves that i listed actually possible. I am very sorry, as you probably expected actual solid proof and all you got was this

anyways here's the updated list:

<details>
  <summary>Click to expand!</summary>
"Ba1b2+", "Ba1b2#", "Ba1c3+", "Ba1c3#", "Ba1d4+", "Ba1d4#", "Ba8b7+", "Ba8b7#", "Ba8c6+", "Ba8c6#", "Ba8d5+", "Ba8d5#", "Bh1e4+", "Bh1e4#", "Bh1f3+", "Bh1f3#", "Bh1g2+", "Bh1g2#", "Bh8e5+", "Bh8e5#", "Bh8f6+", "Bh8f6#", "Bh8g7+", "Bh8g7#", "Qa1h8+", "Qa1h8#", "Qa8h1+", "Qa8h1#", "Qb2a1+", "Qb2a1#", "Qb2h8+", "Qb2h8#", "Qb7a8+", "Qb7a8#", "Qb7h1+", "Qb7h1#", "Qc3a1+", "Qc3a1#", "Qc3h8+", "Qc3h8#", "Qc4a2+", "Qc4a2#", "Qc5a7+", "Qc5a7#", "Qc6a8+", "Qc6a8#", "Qc6h1+", "Qc6h1#", "Qd3b1+", "Qd3b1#", "Qd4a1+", "Qd4a1#", "Qd4b2+", "Qd4b2#", "Qd4h8+", "Qd4h8#", "Qd5a8+", "Qd5a8#", "Qd5b7+", "Qd5b7#", "Qd5h1+", "Qd5h1#", "Qd6b8+", "Qd6b8#", "Qe3g1+", "Qe3g1#", "Qe4a8+", "Qe4a8#", "Qe4g2+", "Qe4g2#", "Qe4h1+", "Qe4h1#", "Qe5a1+", "Qe5a1#", "Qe5g7+", "Qe5g7#", "Qe5h8+", "Qe5h8#", "Qe6g8+", "Qe6g8#", "Qf3a8+", "Qf3a8#", "Qf3h1+", "Qf3h1#", "Qf4h2+", "Qf4h2#", "Qf5h7+", "Qf5h7#", "Qf6a1+", "Qf6a1#", "Qf6h8+", "Qf6h8#", "Qg2a8+", "Qg2a8#", "Qg2h1+", "Qg2h1#", "Qg7a1+", "Qg7a1#", "Qg7h8+", "Qg7h8#", "Qh1a8+", "Qh1a8#", "Qh8a1+", "Qh8a1#", "Qa1xh8+", "Qa1xh8#", "Qa8xh1+", "Qa8xh1#", "Qb2xa1+", "Qb2xa1#", "Qb2xh8+", "Qb2xh8#", "Qb7xa8+", "Qb7xa8#", "Qb7xh1+", "Qb7xh1#", "Qc3xa1+", "Qc3xa1#", "Qc3xh8+", "Qc3xh8#", "Qc6xa8+", "Qc6xa8#", "Qc6xh1+", "Qc6xh1#", "Qd4xa1+", "Qd4xa1#", "Qd4xh8+", "Qd4xh8#", "Qd5xa8+", "Qd5xa8#", "Qd5xh1+", "Qd5xh1#", "Qe4xa8+", "Qe4xa8#", "Qe4xh1+", "Qe4xh1#", "Qe5xa1+", "Qe5xa1#", "Qe5xh8+", "Qe5xh8#", "Qf3xa8+", "Qf3xa8#", "Qf3xh1+", "Qf3xh1#", "Qf6xa1+", "Qf6xa1#", "Qf6xh8+", "Qf6xh8#", "Qg2xa8+", "Qg2xa8#", "Qg2xh1+", "Qg2xh1#", "Qg7xa1+", "Qg7xa1#", "Qg7xh8+", "Qg7xh8#", "Qh1xa8+", "Qh1xa8#", "Qh8xa1+", "Qh8xa1#"
</details>

**Comment by phoenizboi** [comment]
I can at least show you my logic if that helps

**Comment by phoenizboi** [comment]
I discovered that moves like Raa1+ and Raa1# are also impossible, so I added them to my list. Yeah im really sorry and this probably isn't helpful. Im just making a fool of myself at this point, but i'll keep updating my list and posting it here when i inevitably find something new i missed:

<details>
  <summary>Click to expand!</summary>
"Raa1+", "Raa1#", "Raa2+", "Raa2#", "Raa3+", "Raa3#", "Raa4+", "Raa4#", "Raa5+", "Raa5#", "Raa6+", "Raa6#", "Raa7+", "Raa7#", "Raa8+", "Raa8#", "Rah1+", "Rah1#", "Rah2+", "Rah2#", "Rah3+", "Rah3#", "Rah4+", "Rah4#", "Rah5+", "Rah5#", "Rah6+", "Rah6#", "Rah7+", "Rah7#", "Rah8+", "Rah8#", "Rha1+", "Rha1#", "Rha2+", "Rha2#", "Rha3+", "Rha3#", "Rha4+", "Rha4#", "Rha5+", "Rha5#", "Rha6+", "Rha6#", "Rha7+", "Rha7#", "Rha8+", "Rha8#", "Rhh1+", "Rhh1#", "Rhh2+", "Rhh2#", "Rhh3+", "Rhh3#", "Rhh4+", "Rhh4#", "Rhh5+", "Rhh5#", "Rhh6+", "Rhh6#", "Rhh7+", "Rhh7#", "Rhh8+", "Rhh8#", "Ba1b2+", "Ba1b2#", "Ba1c3+", "Ba1c3#", "Ba1d4+", "Ba1d4#", "Ba8b7+", "Ba8b7#", "Ba8c6+", "Ba8c6#", "Ba8d5+", "Ba8d5#", "Bh1e4+", "Bh1e4#", "Bh1f3+", "Bh1f3#", "Bh1g2+", "Bh1g2#", "Bh8e5+", "Bh8e5#", "Bh8f6+", "Bh8f6#", "Bh8g7+", "Bh8g7#", "Qa1h8+", "Qa1h8#", "Qa8h1+", "Qa8h1#", "Qb2a1+", "Qb2a1#", "Qb2h8+", "Qb2h8#", "Qb7a8+", "Qb7a8#", "Qb7h1+", "Qb7h1#", "Qc3a1+", "Qc3a1#", "Qc3h8+", "Qc3h8#", "Qc4a2+", "Qc4a2#", "Qc5a7+", "Qc5a7#", "Qc6a8+", "Qc6a8#", "Qc6h1+", "Qc6h1#", "Qd3b1+", "Qd3b1#", "Qd4a1+", "Qd4a1#", "Qd4b2+", "Qd4b2#", "Qd4h8+", "Qd4h8#", "Qd5a8+", "Qd5a8#", "Qd5b7+", "Qd5b7#", "Qd5h1+", "Qd5h1#", "Qd6b8+", "Qd6b8#", "Qe3g1+", "Qe3g1#", "Qe4a8+", "Qe4a8#", "Qe4g2+", "Qe4g2#", "Qe4h1+", "Qe4h1#", "Qe5a1+", "Qe5a1#", "Qe5g7+", "Qe5g7#", "Qe5h8+", "Qe5h8#", "Qe6g8+", "Qe6g8#", "Qf3a8+", "Qf3a8#", "Qf3h1+", "Qf3h1#", "Qf4h2+", "Qf4h2#", "Qf5h7+", "Qf5h7#", "Qf6a1+", "Qf6a1#", "Qf6h8+", "Qf6h8#", "Qg2a8+", "Qg2a8#", "Qg2h1+", "Qg2h1#", "Qg7a1+", "Qg7a1#", "Qg7h8+", "Qg7h8#", "Qh1a8+", "Qh1a8#", "Qh8a1+", "Qh8a1#", 
"Raxa1+", "Raxa1#", "Raxa8+", "Raxa8#", "Raxh1+", "Raxh1#", "Raxh8+", "Raxh8#", "Rhxa1+", "Rhxa1#", "Rhxa8+", "Rhxa8#", "Rhxh1+", "Rhxh1#", "Rhxh8+", "Rhxh8#", "Qa1xh8+", "Qa1xh8#", "Qa8xh1+", "Qa8xh1#", "Qb2xa1+", "Qb2xa1#", "Qb2xh8+", "Qb2xh8#", "Qb7xa8+", "Qb7xa8#", "Qb7xh1+", "Qb7xh1#", "Qc3xa1+", "Qc3xa1#", "Qc3xh8+", "Qc3xh8#", "Qc6xa8+", "Qc6xa8#", "Qc6xh1+", "Qc6xh1#", "Qd4xa1+", "Qd4xa1#", "Qd4xh8+", "Qd4xh8#", "Qd5xa8+", "Qd5xa8#", "Qd5xh1+", "Qd5xh1#", "Qe4xa8+", "Qe4xa8#", "Qe4xh1+", "Qe4xh1#", "Qe5xa1+", "Qe5xa1#", "Qe5xh8+", "Qe5xh8#", "Qf3xa8+", "Qf3xa8#", "Qf3xh1+", "Qf3xh1#", "Qf6xa1+", "Qf6xa1#", "Qf6xh8+", "Qf6xh8#", "Qg2xa8+", "Qg2xa8#", "Qg2xh1+", "Qg2xh1#", "Qg7xa1+", "Qg7xa1#", "Qg7xh8+", "Qg7xh8#", "Qh1xa8+", "Qh1xa8#", "Qh8xa1+", "Qh8xa1#"
</details>

**Comment by jacksonthall22** [comment]
Wow, that's very diligent of you, but unfortunately I will need code/pseudocode example to make changes. I just haven't had time to get back into this repo yet, but it should be something along the lines of:

1. Start with the generated list of SANs without any + or # symbols - I am confident the logic for these is correct now
2. Loop through these strings and extract the piece type, destination square, whether it's a capture, and the file and/or rank disambiguators (if any)

Then either:
3a. Brute force solution (easy):
	a. Generate all possible piece configurations with pairs/triplets of that piece type that would cause the SAN to require the observed disambiguation
	b. Check if, in any of these configurations, making the given SAN move causes new squares to be attacked (note that a capture could make this true instead of false, at least for file-disambiguated moves, ex. `Baxc3` as opposed to `Bac3` with bishops on `a1` and `e1`). If so, we need to add this SAN with both + and # appended.

or:

3b. Logical solution (harder):
	a. Come up with some logic to understand what must be true of any configuration causing the SAN to require the observed disambiguation. My intuition (which can't be trusted) says any configuration requiring the observed disambiguation is "equivalent" to any other (if any configuration causes new squares to be attacked, they all will - this will either need solid logic to be convincing or a counterexample). If that's the case, generate a single configuration that causes this disambiguation.
	b. Check if making the given SAN move in this configuration causes new squares to be attacked. If so, add this SAN with both + and # appended.

**Comment by phoenizboi** [comment]
I tried to come up with some sort of logic, it's not very good but it's the best I have for now.

This issue happens when the disambiguations force the pieces into an orientation where no new pieces are attacked by the move. For a capture to also have this problem, I found that the only way is for the move to land on a corner, which will become obvious as I explain.

For pawns:
<details>
  <summary>Click to expand!</summary>
Pawns can only be single disambiguated by file, and for them to be disambiguated, a capture is required. Pawns only attack the squares on the rank one ahead of them. Every time a pawn moves, it moves forward one rank, so a pawn will always attack new squares. Promotion will always attack new squares. Therefore, disambiguation for pawns can always result in a check or checkmate.

![IMG_0610](https://github.com/user-attachments/assets/b1d26b01-fc26-40b7-8692-fba958fa714a)
</details>

For knights:
<details>
  <summary>Click to expand!</summary>
For a knight to not attack any new squares, they must all be either off of the board or occupied by another knight. Double disambiguation requires the knights to be arranged in a 5x3 rectangle, which forces the destination square to not be on the edge of the board, allowing for newly attacked squares:

![IMG_0611](https://github.com/user-attachments/assets/0fbae532-56a5-4c48-88ab-bc9731fed176)

Single disambiguation is a little promising, because the corners block all but two squares:

![IMG_0612](https://github.com/user-attachments/assets/460836c4-e695-458f-a826-f0cc326b8c5a)

However, this is easily countered with a discovered check:

![IMG_0613](https://github.com/user-attachments/assets/97c85af8-db31-4ad5-9e02-5065e1c0252e)

So all the cases are covered: knight moves can always cause check or checkmate.

The other pieces (rooks, bishops, and queens) can have this issue. However, captures must be treated seperately.
</details>

Rooks:
<details>
  <summary>Click to expand!</summary>
Rooks have this problem every time they start on an edge AND both rooks are on different ranks and files. they don't start on the edge of the board, they can reveal a discovered check:

![IMG_0614](https://github.com/user-attachments/assets/989b1a6e-a992-4b00-b1f4-e3ac0b7c9903)


If the rooks can occupy the same rank or file, they can attack new squares:

![IMG_0615](https://github.com/user-attachments/assets/75d4ca01-9677-483d-868c-25a49a238de8)


But if they start on an edge, and they are forced to lie on different ranks and files, then the other rook already covers the squares that the moved rook tries to attack:

![IMG_0616](https://github.com/user-attachments/assets/ece8c129-a146-4139-8482-f0e8c1b61fea)
</details>

Edit: so while making this and trying to figure out what exactly can cause both issues, i found more moves that don't work, and i must extend my list again :(

Rooks:
<details>
  <summary>Click to expand!</summary>
Rank disambiguation never causes this issue as it lets the rooks lie on the same file.

The move must end on the a or h file because otherwise the other rook can lie on the same rank.

So you can just check for both requirements: if both rooks MUST lie on different ranks and files, and if the moving rook MUST start on an edge.

For captures: the same restrictions apply, but the move must also end on a corner. Otherwise, a piece can be captured to reveal a check:

![IMG_0617](https://github.com/user-attachments/assets/db31da6e-38f1-470b-8a17-96e6d105ab56)
</details>

Bishops:
<details>
  <summary>Click to expand!</summary>
The move must start on a corner to have this problem. That's it. 

![IMG_0618](https://github.com/user-attachments/assets/8d57be00-ebe5-46db-b252-ff770d473992)


For this to happen, the move must be doubly disambiguated. If it isn't, it allows for the bishop to not start on the corner:

![IMG_0619](https://github.com/user-attachments/assets/701e9408-c343-40b0-a5cf-98ee34da9119)


And if it doesn't start on a corner, then discovered check is always possible.

Also, none of the bishop captures have this problem, because they can just capture and reveal a check:

![IMG_0627](https://github.com/user-attachments/assets/ca8922c7-2d6a-464d-b780-3c2d5a3999d6)


So, simply search for all the bishop moves that start in the corner and don't capture a piece.
</details>

Queens:
<details>
  <summary>Click to expand!</summary>
There are a couple ways this can happen, but both ways require double disambiguation. The simplest is if the queen moves diagonally into a corner:

![IMG_0620](https://github.com/user-attachments/assets/253364e0-690c-40d1-afab-0b937bab571f)


The edge of the board forces the other queens to be in spots that block/cover the moved queen. The diagonal is always attacked. Also, the corner also blocks potential new directions the queen would attack. Here's an example of how the queen could attack new squares if it doesn't land on a corner:

![IMG_0621](https://github.com/user-attachments/assets/8c668d9b-06fb-45dc-948d-4697bccf5760)


The blue lines are always covered, but the red lines that were blocked by the corner are now exposed. Therefore, it must land on a corner, except for ONE SINGLE random edge case I found (but for a good reason):

![IMG_0622](https://github.com/user-attachments/assets/0f1b2de6-a998-4dfc-854a-b6e90a71c5db)


The other queens just so happen to cover the exposed red lines. If they were one square further away from the corner, the queen could move to a spot that doesn't block the moved queen:

![IMG_0623](https://github.com/user-attachments/assets/4ef28528-0aff-4220-ba05-48e9711f7b74)


And if the queens were further apart, they don't cover the red lines:

![IMG_0621](https://github.com/user-attachments/assets/76e97122-2993-482c-a44f-c1e2f7a1840e)


Also, this square can be shifted around a bit:

![IMG_0624](https://github.com/user-attachments/assets/5a22f26b-b96e-43a7-8331-23662857b51b)
![IMG_0625](https://github.com/user-attachments/assets/68392796-14b3-45e3-9ec0-9a5d3cc0dc5a)



If it is fully pushed against the corner, that's just the first case of the queen landing in the corner. For captures, most of the cases still apply, except for the outlier that i just described. Because the queen no longer lands on the corner, it can reveal a check by capturing a piece:

![IMG_0626](https://github.com/user-attachments/assets/cbb5fd8b-f49d-461b-bc9e-3e2bd2f3d0ad)


Queens are a little complicated, and I can't really come up with a rigid proof for these. However, the basic idea is that single disambiguation gives too much wiggle room, and the edge of the board must be used to force the queens into awkward positions that block the moving queen.
</details>

**Comment by phoenizboi** [comment]
I'm just gonna paste the ones that I missed, instead of everything again. I'm almost positive that this is everything, because I solidified my logic, but then again I have made changes like 5 times so

<details>
  <summary>Click to expand!</summary>
"Rbxa1+", "Rbxa1#", "Rbxa8+", "Rbxa8#", "Rbxh1+", "Rbxh1#", "Rbxh8+", "Rbxh8#", "Rcxa1+", "Rcxa1#", "Rcxa8+", "Rcxa8#", "Rcxh1+", "Rcxh1#", "Rcxh8+", "Rcxh8#", "Rdxa1+", "Rdxa1#", "Rdxa8+", "Rdxa8#", "Rdxh1+", "Rdxh1#", "Rdxh8+", "Rdxh8#", "Rexa1+", "Rexa1#", "Rexa8+", "Rexa8#", "Rexh1+", "Rexh1#", "Rexh8+", "Rexh8#", "Rfxa1+", "Rfxa1#", "Rfxa8+", "Rfxa8#", "Rfxh1+", "Rfxh1#", "Rfxh8+", "Rfxh8#", "Rgxa1+", "Rgxa1#", "Rgxa8+", "Rgxa8#", "Rgxh1+", "Rgxh1#", "Rgxh8+", "Rgxh8#"
</details>

**Comment by AXZIOM11** [comment]
Wow this thread is really cool, I also stumbled into this problem recently and made my own list and found this after trying to find anyone else who has done this, here is mine:

<details>
  <summary>Click to expand!</summary>
Total {372}

Bishop {24}
Check/Checkmate {24}
Ba1b2+, Ba1b2#, Ba1c3+, Ba1c3#, Ba1d4+, Ba1d4# Ba8b7+, Ba8b7#, Ba8c6+, Ba8c6#, Ba8d5+, Ba8d5#, Bh1e4+, Bh1e4#, Bh1f3+, Bh1f3#, Bh1g2+, Bh1g2#, Bh8e5+, Bh8e5#, Bh8f6+, Bh8f6#, Bh8g7+, Bh8g7#,

Rook {176}
Capture/Check/Checkmate {128}
Raa1+, Raa1#, Raxa1+, Raxa1#, Raa8+, Raa8#, Raxa8+, Raxa8#, Rah1+, Rah1#, Raxh1+, Raxh1#, Rah8+, Rah8#, Raxh8+, Raxh8#, Rba1+, Rba1#, Rbxa1+, Rbxa1#, Rba8+, Rba8#, Rbxa8+, Rbxa8#, Rbh1+, Rbh1#, Rbxh1+, Rbxh1#, Rbh8+, Rbh8#, Rbxh8+, Rbxh8#, Rca1+, Rca1#, Rcxa1+, Rcxa1#, Rca8+, Rca8#, Rcxa8+, Rcxa8#, Rch1+, Rch1#, Rcxh1+, Rcxh1#, Rch8+, Rch8#, Rcxh8+, Rcxh8#, Rda1+, Rda1#, Rdxa1+, Rdxa1#, Rda8+, Rda8#, Rdxa8+, Rdxa8#, Rdh1+, Rdh1#, Rdxh1+, Rdxh1#, Rdh8+, Rdh8#, Rdxh8+, Rdxh8#, Rea1+, Rea1#, Rexa1+, Rexa1#, Rea8+, Rea8#, Rexa8+, Rexa8#, Reh1+, Reh1#, Rexh1+, Rexh1#, Reh8+, Reh8#, Rexh8+, Rexh8#, Rfa1+, Rfa1#, Rfxa1+, Rfxa1#, Rfa8+, Rfa8#, Rfxa8+, Rfxa8#, Rfh1+, Rfh1#, Rfxh1+, Rfxh1#, Rfh8+, Rfh8#, Rfxh8+, Rfxh8#, Rga1+, Rga1#, Rgxa1+, Rgxa1#, Rga8+, Rga8#, Rgxa8+, Rgxa8#, Rgh1+, Rgh1#, Rgxh1+, Rgxh1#, Rgh8+, Rgh8#, Rgxh8+, Rgxh8#, Rha1+, Rha1#, Rhxa1+, Rhxa1#, Rha8+, Rha8#, Rhxa8+, Rhxa8#, Rhh1+, Rhh1#, Rhxh1+, Rhxh1#, Rhh8+, Rhh8#, Rhxh8+, Rhxh8#, 
Check/Checkmate {48}
Raa2+, Raa2#, Raa3+, Raa3#, Raa4+, Raa4#, Raa5+, Raa5#, Raa6+, Raa6#, Raa7+, Raa7#, Rah2+, Rah2#, Rah3+, Rah3#, Rah4+, Rah4#, Rah5+, Rah5#, Rah6+, Rah6#, Rah7+, Rah7#, Rha2+, Rha2#, Rha3+, Rha3#, Rha4+, Rha4#, Rha5+, Rha5#, Rha6+, Rha6#, Rha7+, Rha7#, Rhh2+, Rhh2#, Rhh3+, Rhh3#, Rhh4+, Rhh4#, Rhh5+, Rhh5#, Rhh6+, Rhh6#, Rhh7+, Rhh7#, 

Queen {172}
Capture/Check/Checkmate {112}
Qa1h8+, Qa1h8#, Qa1xh8+, Qa1xh8#, Qa8h1+, Qa8h1#, Qa8xh1+, Qa8xh1#, Qb2a1+, Qb2a1#, Qb2xa1+, Qb2xa1#, Qb2h8+, Qb2h8#, Qb2xh8+, Qb2xh8#, Qb7a8+, Qb7a8#, Qb7xa8+, Qb7xa8#, Qb7h1+, Qb7h1#, Qb7xh1+, Qb7xh1#, Qc3a1+, Qc3a1#, Qc3xa1+, Qc3xa1#, Qc3h8+, Qc3h8#, Qc3xh8+, Qc3xh8#, Qc6a8+, Qc6a8#, Qc6xa8+, Qc6xa8#, Qc6h1+, Qc6h1#, Qc6xh1+, Qc6xh1#, Qd4a1+, Qd4a1#, Qd4xa1+, Qd4xa1#, Qd4h8+, Qd4h8#, Qd4xh8+, Qd4xh8#, Qd5a8+, Qd5a8#, Qd5xa8+, Qd5xa8#, Qd5h1+, Qd5h1#, Qd5xh1+, Qd5xh1#, Qe4a8+, Qe4a8#, Qe4xa8+, Qe4xa8#, Qe4h1+, Qe4h1#, Qe4xh1+, Qe4xh1#, Qe5a1+, Qe5a1#, Qe5xa1+, Qe5xa1#, Qe5h8+, Qe5h8#, Qe5xh8+, Qe5xh8#, Qf3a8+, Qf3a8#, Qf3xa8+, Qf3xa8#, Qf3h1+, Qf3h1#, Qf3xh1+, Qf3xh1#, Qf6a1+, Qf6a1#, Qf6xa1+, Qf6xa1#, Qf6h8+, Qf6h8#, Qf6xh8+, Qf6xh8#, Qg2a8+, Qg2a8#, Qg2xa8+, Qg2xa8#, Qg2h1+, Qg2h1#, Qg2xh1+, Qg2xh1#, Qg7a1+, Qg7a1#, Qg7xa1+, Qg7xa1#, Qg7h8+, Qg7h8#, Qg7xh8+, Qg7xh8#, Qh1a8+, Qh1a8#, Qh1xa8+, Qh1xa8#, Qh8a1+, Qh8a1#, Qh8xa1+, Qh8xa1#, 
Check/Checkmate {24}
Qc4a2#, Qc4a2+, Qc5a7#, Qc5a7+, Qd3b1#, Qd3b1+, Qd4b2#, Qd4b2+, Qd5b7#, Qd5b7+, Qd6b8#, Qd6b8+, Qe3g1#, Qe3g1+, Qe4g2#, Qe4g2+, Qe5g7#, Qe5g7+, Qe6g8#, Qe6g8+, Qf4h2#, Qf4h2+, Qf5h7#, Qf5h7+, 
Capture/Check {24}
Qd4a7+, Qd4g1+, Qd4g7+, Qd4xa7+, Qd4xg1+, Qd4xg7+, Qd5a2+, Qd5g2+, Qd5g8+, Qd5xa2+, Qd5xg2+, Qd5xg8+, Qe4b1+, Qe4b7+, Qe4h7+, Qe4xb1+, Qe4xb7+, Qe4xh7+, Qe5b2+, Qe5b8+, Qe5h2+, Qe5xb2+, Qe5xb8+, Qe5xh2+, 
Capture+Check {12}
Qc4xa2+, Qc5xa7+, Qd3xb1+, Qd4xb2+, Qd5xb7+, Qd6xb8+, Qe3xg1+, Qe4xg2+, Qe5xg7+, Qe6xg8+, Qf4xh2+, Qf5xh7+, 
</details>

**Comment by AXZIOM11** [comment]
Notably there is moves like Qd4a7+ or Qc4xa2+ which can be checkmate but not check which is really crazy that it can even happen, and you also missed some rook moves like Rba1#. I did all of this manually so there's always a chance that I missed stuff or made a typo somewhere, but if this is right then there is 28902 valid chess notations.

**Comment by phoenizboi** [comment]
> Notably there is moves like Qd4a7+ or Qc4xa2+ which can be checkmate but not check which is really crazy that it can even happen, and you also missed some rook moves like Rba1#. I did all of this manually so there's always a chance that I missed stuff or made a typo somewhere, but if this is right then there is 28902 valid chess notations.

I'm so sorry, but I stumbled upon this too. I eventually discovered that it CAN be check, if there is a piece of the opposite color that can capture the checking queen (which is the only way for it to not be checkmate). 

By the way, I did miss some! Congrats for noticing! I just gave up adding to this considering the lack of activity, but I continued to work on it by myself for reasons that will take me way too long to explain

Here is the raw notepad file that I had saved in my phone for a while, lemme just dust it off

[impossible san things.txt](https://github.com/user-attachments/files/23148264/impossible.san.things.txt)

According to my count, there are 28,938 valid chess notations. Please check my work, I did it a very long time ago (and I waited so long for someone to look at this lol). I might try and prove this is correct at some point

**Comment by phoenizboi** [comment]
At some point I even made a puzzle where the only drawing move was one of the moves I once thought was impossible!

<img width="1200" height="630" alt="Image" src="https://github.com/user-attachments/assets/79117b9a-ddab-4a6f-ba88-69982b0836bc" />

**Comment by AXZIOM11** [comment]
Oh wow that's so cool you are right, I forgot you can do that, and yeah this problem has been like a worm in my brain ever since someone asked me "is there moves in chess that cannot be checkmate" unassumingly. Your list seems like the same as mine now with the check only move removed, so I'm not sure what else we could be both missing, and idk how to program to prove anything sadly.

**Comment by jacksonthall22** [comment]
Fun to see this issue getting attention 😄 this problem has also been a worm in my brain and deserves to be solved. I have a bit of free time so maybe I'll take another stab at the logic here.

But first, since this repo has gotten a few stars, I think it would be nice to be confident that the list here is correct and make it a single source of truth for this common question, so I'm trying to write some efficient code to brute-force the full list. Will update soon!

I made some edits above for brevity, let's use this markdown feature in the future:

```html
<details>
  <summary>Click to expand!</summary>

  Long text here!
</details>
```

<details>
  <summary>Click to expand!</summary>

  Long text here!
</details>

**Comment by jacksonthall22** [comment]
Hmm, an accurate brute force of this problem might be completely intractable because we'd have to construct tons of positions to verify whether some checkmates may be possible for each disambiguated SAN (prove me wrong?). Just going to drop some observations/summarization on previous discussions.

## What the code already does well
I recently updated [`hyprchs/chess_action_space`](https://github.com/hyprchs/chess_action_space), which is giving me some more ideas about this problem. The minimal action space in chess is basically the space of all valid UCI move strings (with a slight caveat that queen promotion moves can be coalesced to the corresponding non-promotion UCI to make the space minimal - ie. assumes the user can assume that an unspecified promotion piece should mean a queen, so `e7e8q` is represented as `e7e8`).

Actions are sort of "atomic" and maybe helpful when thinking about this problem because every action can turn into many SANs given different board states. The code in this repo already has robust logic for answering the following questions:

Given an action and a piece type,
- can the move require a file disambiguator?
- can the move require a rank disambiguator?
- can the move require a full-square disambiguator?

## What this issue is about
This issue is about fixing the way the code answers these follow-up questions:

Given a non-check, non-checkmate SAN move *that we've proven can exist above* (and may or may not have disambiguators),
1. can that SAN move also cause check?
2. can that SAN move also cause checkmate?

Rereading everything above, I realized it's incredibly relevant that in chess in general, check can only be caused in 2 (really 3) ways:
A. The moving piece checks the king
B. The moving piece reveals a discovered check
(C. Castling causes the involved rook to check the enemy king, but it's not really relevant here because we can trivially prove with examples that `O-O+`/`O-O#`/`O-O-O+`/`O-O-O#` are all possible.)

We can be confident about the answer to question `1` above if we can reliably determine whether both `A` and `B` can *ever* be true for some base SAN move. If neither can ever be true, then that SAN cannot cause check nor checkmate.

Some corollaries about this:
- More of a side note: determining whether these conditions are true for the SAN's corresponding *action* as opposed to the SAN itself does not help determine whether the SAN can cause check/checkmate, because if there are disambiguators in the specific SAN, they impose additional restrictions on the board state that might change the answer (again, the SAN may be one of many possible SANs that correspond to the same underlying action).
- If we assume we are only interested in generating SANs that come from legal positions (we are here), then `A` is equivalent to the condition that the SAN can cause new squares to be attacked, related to the observation as originally stated in the issue (otherwise the king would have already been in check - a contradiction).
  - Note the phrasing "can cause" as opposed to "causes". A base SAN string does not necessarily specify exactly one piece configuration of the moving piece type - rather, disambiguators can specify some *set* of configurations, given the *piece type* and *action*, that would result in the observed disambiguation being required in the SAN. The question would then be whether *any* possible configuration of pieces for that SAN causes new squares to be attacked. I'm not convinced that's an easy question to solve in a robust "proof" given that there are multiple piece types and multiple actions (the ideal algorithm/argument would be agnostic to the piece type and action, and would be based entirely on board geometry and the "slider type" of the moving piece - horizontal/vertical or diagonal - and probably special cases for knights/pawns).
  - @phoenizboi, you insightfully noted that some SANs for captures may cause new squares to be attacked when the non-capture version of the SAN would not (🤯). I think this fact only is relevant to the behavior of full-square disambiguated moves, but in any case, it will be tractable to determine exactly when this happens with brute force.
  - Side note that it's not an equivalent statement to say that `A` would be violated if the SAN move fails to cause previously-attacked squares to become un-attacked. We just care about newly attacked squares because those are squares the enemy king could stand in order for the SAN to deliver check and/or checkmate. I think we can use brute force over that set of newly-attacked squares and validate, given the moving piece type, that there are no cases where checkmate is possible but check is impossible (my intuition says this is the case), and somehow convince ourselves that there are no cases where check is possible but checkmate is impossible (I have no idea how to even attempt to prove that, but I can't think of any counterexamples).
- `B` also implies some additional constraints. A discovered check occurring implies two things: `b1.` the moving piece was not already checking the king, and `b2.` there is a slider (bishop, rook, queen) behind the moving piece that is making the check. Therefore, a general rule is that if the moving piece is a slider, then it can only cause a discovered check when there is a slider behind it that can move along a direction that the moving slider piece does *not* move (either horizontal/vertical or diagonal) - otherwise one of those points would be violated. Rooks are horizontal/vertical sliders, so they can only cause discovered checks from diagonal sliders; bishops are diagonal sliders, so they can only cause discovered checks from horizontal/vertical sliders; queens slide in all directions, so they cannot cause any discovered checks. This means that when the moving piece is a horizontal/vertical slider (rook/queen) and starts from an edge of the board, no *diagonal* discovered checks are possible (it would require either the diagonal checker or the enemy king to be off the board); therefore it cannot cause *any* discoveries. The same goes when the moving piece is a diagonal slider (bishop/queen) and starts from a corner; no *horizontal/vertical* discovered checks are possible (they would require either the horizontal/vertical checker or the enemy king to be off the board); therefore it also cannot cause *any* discoveries. Queens inherit the properties of both the bishop and rook here, so a queen moving either horizontally/vertically or diagonally (doesn't matter) from any board edge (or any corner, to be complete, but these are obviously included among the edge squares) cannot cause discovered checks.

Some edge cases above are maybe better generalized by this logic now:
- `Ba1c3` fails on `A` because the full-square disambiguator implies the presence of 3 bishops on `a1`, `a3`, and `c1`, where `Ba1c3` wouldn't cause new squares to be attacked. It also fails on `B` because it comes from a corner (diagonal sliders coming from a corner cannot make discovered checks) 
- `Rae4` fails on `A` because this SAN cannot cause new squares to be attacked (e file was already covered by another rook, and 4th rank remains attacked). It also fails on `B` because it comes from an edge (horizontal/vertical sliders coming from an edge cannot make discovered checks).

Open question that we would need to answer for any insights gleaned above to be useful (tangential to your question @AXZIOM11): Are there any SAN moves that can cause check but cannot cause checkmate?

I know most of this is just stating the obvious, but I think it's important to do that for problems like these - start by saying things we know, keep saying things we know, eventually reach the answer? 👀

**Comment by AXZIOM11** [comment]
> Are there any SAN moves that can cause check but cannot cause checkmate?

The answer is no because if there is a check on a king then regardless of where anything is you can block any tiles around the king with bishops and rooks to make it checkmate, among many other methods. I also think this works the other way around because a a tile can be attacked/defended by at minimum 5 directions, and double disambiguations can only ever eliminate 3 of those 5+ directions. Combined with the fact that a double check isn't ever needed, ergo all checks/checkmates are an attack from 1 tile, the piece that is attacking the king can always be defended and therefore every checkmate can also be check. The logic there is a bit harder to make rigorous but I think it's right.

I have been thinking about how a computer would brute force this problem, as its really easy to prove that a notation is possible but much harder to prove that it's impossible. Here is my idea so far: For every notation n without check or checkmate, check every possible move that can lead to notation n, checking each one of those 9 times for a different type of discovered attack angle from the moved piece, checking each time for if there are any new tiles that can be attacked. If there are no positions that works then its impossible and if there is you can save the position and move to the next notation as with the logic mentioned above it should be possible to covert a check to a checkmate and vice versa. I'm pretty sure this is sufficient to prove that a position is impossible but I don't know how to make that rigorous. You can also skip the all of the discovered checks step with since they already cover every direction, and half of the step for bishop and rooks for the same reason. I would try implementing this myself but I am learning to program so I don't know how yet.

**Comment by phoenizboi** [comment]
>I know most of this is just stating the obvious, but I think it's important to do that for problems like these - start by saying things we know, keep saying things we know, eventually reach the answer? 👀

This may be a little too obvious, but for queen moves, can we simply create a list of every square attacked by each of the 3 queens, then create a second list of every square attacked by the moved queen's new square, and check if there is at least one square in the second list that is not contained in the first list?

This logic would still apply for captures, it just would be a bit more complicated to find which squares would be attacked.

Also, we would need to do this for every possible configuration of queens that result in the same double disambiguation move, which might be hard. However, we would just need to find one configuration that satisfies the condition. If none of the configurations satisfy the condition, then the move cannot cause check/checkmate.

Also, it feels like it should be obvious that there would always be a way to capture the checking piece, implying that check is always possible when checkmate is, but it feels like there is no way to prove it without brute force. It's easy to say that "there are enough places on the board that we can find somewhere to put an extra piece" but it's not that simple.

**Comment by jacksonthall22** [comment]
> a tile can be attacked/defended by at minimum 5 directions, and double disambiguations can only ever eliminate 3 of those 5+ directions

can you elaborate @AXZIOM11? Are you thinking of the 5 directions being like this?

<details>
  <summary>Click to expand!</summary>

  <img width="325" height="324" alt="Image" src="https://github.com/user-attachments/assets/f5ac5677-30be-4dc8-92dd-8ad4d3dcc0d9" />
</details>

Wondering if this idea could be developed and actually help with a logical proof. Again, I can't think of any example, but I was vaguely worried that the geometry required by ex. a full-square disambiguated SAN could require that the other pieces be arranged in such a way that you can never block every escape square for the king, meaning check would be possible but checkmate would not be (I would be very surprised if this were possible geometrically but I don't know how to disprove it).

> for queen moves, can we simply create a list of every square attacked by each of the 3 queens, then create a second list of every square attacked by the moved queen's new square, and check if there is at least one square in the second list that is not contained in the first list?

Yeah, that's exactly what I was thinking here:

> I think we can use brute force over that set of newly-attacked squares and validate, given the moving piece type, that there are no cases where checkmate is possible but check is impossible (my intuition says this is the case)

I think this should not be too hard to brute force.

I'm working on restructuring the code right now to be more modular and generally easier to follow, which I think will help show/test that every individual part is working correctly: #5

For provably-possible SANs, a brute-force solution basically means we must be able to generate a FEN where `python-chess` shows that exact string when parsing the move against that position.

It's definitely harder to prove invalid ones, but it would be nice to be able to generate some kind of debug image ([`hyprchs/web-boardimage`](https://github.com/hyprchs/web-boardimage) forked from niklasf) that shows with arrows why why it's impossible, ex. due to board geometry or because it violates one of the conditions `A` or `B` in my previous comment.

**Comment by AXZIOM11** [comment]
> Are you thinking of the 5 directions being like this?

Yeah that image is right, I was trying to describe the idea that there are always more ways to defend a square than there is to attack a square via notations.

For the boardimage thing you are mentioning, would the rook moves have 7 images or 1 image? Because for those moves it doesn't specify where 1 of the 2 rooks is, which means it can be in 7 different positions, unlike all the other moves so far which always only have 1 possible way for the pieces to be placed.
